ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Node/DiffusionPipelining.hs

```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeOperators       #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Ouroboros.Consensus.HardFork.Combinator.Node.DiffusionPipelining () where
+
+import           Data.Functor.Product
+import           Data.Proxy
+import           Data.SOP.BasicFunctors
+import qualified Data.SOP.Match as Match
+import           Data.SOP.NonEmpty
+import           Data.SOP.Strict
+import           Ouroboros.Consensus.Block.SupportsDiffusionPipelining
+import           Ouroboros.Consensus.HardFork.Combinator.Abstract
+import           Ouroboros.Consensus.HardFork.Combinator.AcrossEras
+import           Ouroboros.Consensus.HardFork.Combinator.Basics
+import           Ouroboros.Consensus.HardFork.Combinator.Block
+import           Ouroboros.Consensus.TypeFamilyWrappers
+
+-- | The 'BlockSupportsDiffusionPipelining' instance for the HFC is
+-- compositional:
+--
+--  - @'TentativeHeaderState' ('HardForkBlock' xs)@ is
+--    'OneEraTentativeHeaderState', the n-ary sum of per-era
+--    'TentativeHeaderState's. The 'initialTentativeHeaderState' is the one for
+--    the first era.
+--
+--  - @'updateTentativeHeaderState' bcfg hdr st@ works like this:
+--
+--      - If @hdr@ and @st@ are in the same era, we call the corresponding
+--        era-specific 'updateTentativeHeaderState'.
+--
+--      - If @hdr@ is in a later era than @st@, we replace @st@ with the
+--        'initialTentativeHeaderState' for the era of @hdr@, and then proceed
+--        like in the previous step.
+--
+--      - If @hdr@ is in an earlier era than @st@, we return 'Nothing' to
```

Comment:
Maybe call out that this has some undesirable consequences, but they don't seem worth the extra complexity?